azure-iot-device/azure/iot/device/iothub/pipeline/mqtt_pipeline.py

```diff
@@ -155,26 +156,66 @@ def _on_disconnected():
             if self.on_disconnected:
                 self.on_disconnected()
 
+        # Set internal event handlers
         self._pipeline.on_pipeline_event_handler = _on_pipeline_event
         self._pipeline.on_connected_handler = _on_connected
         self._pipeline.on_disconnected_handler = _on_disconnected
 
+        # Initialize the pipeline
         callback = EventedCallback()
-
-        # NOTE: It would be nice if this didn't have to go down as a dynamic operation.
-        # At this time we haven't been able to figure out a better way to make it work though.
-
         op = pipeline_ops_base.InitializePipelineOperation(callback=callback)
-
         self._pipeline.run_op(op)
         callback.wait_for_completion()
 
+        # Set the running flag
+        self._running = True
+
+    def _ensure_running(self):
+        if not self._running:
+            raise pipeline_exceptions.PipelineNotRunning(
+                "Cannot execute method - Pipeline is not running"
+            )
+
+    def shutdown(self, callback):
+        """Shut down the pipeline and clean up any resources.
+
+        Once shut down, making any further calls on the pipeline will result in a
+        PipelineNotRunning exception being raised.
+
+        There is currently no way to resume pipeline functionality once shutdown has occurred.
+
+        :raises: :class:`azure.iot.device.iothub.pipeline.exceptions.PipelineNotRunning` if the
+            pipeline has already been shut down
+
+        The shutdown process itself is not expected to fail under any normal condition, but if it
+        does, exceptions are not "raised", but rather, returned via the "error" parameter when
+        invoking "callback".
+        """
+        self._ensure_running()
```

Comment:
>self._ensure_running() [](start = 8, length = 22)

I would assume that `shutdown` is idempotent.  In other words, if you call `shutdown` and you're already shutdown, nothing needs to happen and the function can complete successfully.  Other functions at this level behave this way -- if you call `connect` and you're already connected, it's not a failure.

Reply:
The idea here is that if you try to shutdown once you are already shut down, it will raise an error. While I could have just let it automatically succeed, I felt it made more sense to raise an error, so that there's a consistent behavior - once you shut down, every future call will fail.